pkg/syncer/upsync/upsync_controller.go

```diff
@@ -0,0 +1,450 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upsync
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	kcpdynamic "github.com/kcp-dev/client-go/dynamic"
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	ddsif "github.com/kcp-dev/kcp/pkg/informer"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	syncerindexers "github.com/kcp-dev/kcp/pkg/syncer/indexers"
+	"github.com/kcp-dev/kcp/pkg/syncer/shared"
+)
+
+const controllerName = "kcp-resource-upsyncer"
+
+var namespaceGVR schema.GroupVersionResource = corev1.SchemeGroupVersion.WithResource("namespaces")
+
+// NewUpSyncer returns a new controller which upsyncs, through the Usyncer virtual workspace, downstream resources
+// which are part of the upsyncable resource types (fixed limited list for now), and provide
```

Comment:
**question:** What are the upsyncable resource types?

Reply:
I _think_ upsyncable resources are limited to `pods`, `endpoints` and `persistentvolumeclaims`. Both experimentation and code support my understanding:

https://github.com/bipuladh/kcp/blob/6cf3293ff35c3558e032c0cd3e4a00724c00e9ec/pkg/syncer/syncer.go#L236-L245

Reply:
That's right, for now only these 3 resources are upsynced, since they correspond to the use-cases forseeable short-term.
Mid-term / long-term, we should have followup work that would allow customizing this list at the KCP level (and possibly storing it to the Synctarget status), but this still needs further design and discussion, especially to keep security in mind.

We should probably open a dedicated issue for this, to track the first thoughts we had in the regard, and further discuss the details.

Reply:
The upsync controller is agnostic which GVRs are upsynced. But we are very opinionated in the wiring into the syncer for the moment. Whether we open that up eventually in some way, we will see.

Reply:
Thanks @davidfestal @sttts 

> We should probably open a dedicated issue for this, to track the first thoughts we had in the regard, and further discuss the details.

I can take care of that.

Reply:
I created the related issue here: https://github.com/kcp-dev/kcp/issues/2804

Let's discuss in this new issue.

Reply:
Thanks ! I will copy the Crossplane use case description there.

Reply:
Thanks for creating the issue @davidfestal you were faster than me 🙇🏻